python/src/main/java/org/apache/zeppelin/python/IPythonInterpreter.java

```diff
@@ -0,0 +1,276 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package org.apache.zeppelin.python;
+
+import org.apache.commons.exec.CommandLine;
+import org.apache.commons.exec.DefaultExecutor;
+import org.apache.commons.exec.ExecuteException;
+import org.apache.commons.exec.ExecuteResultHandler;
+import org.apache.commons.exec.ExecuteWatchdog;
+import org.apache.commons.exec.LogOutputStream;
+import org.apache.commons.exec.PumpStreamHandler;
+import org.apache.commons.exec.environment.EnvironmentUtils;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.zeppelin.interpreter.Interpreter;
+import org.apache.zeppelin.interpreter.InterpreterContext;
+import org.apache.zeppelin.interpreter.InterpreterResult;
+import org.apache.zeppelin.interpreter.remote.RemoteInterpreterUtils;
+import org.apache.zeppelin.interpreter.thrift.InterpreterCompletion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import py4j.GatewayServer;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * IPython Interpreter for Zeppelin
+ */
+public class IPythonInterpreter extends Interpreter implements ExecuteResultHandler {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IPythonInterpreter.class);
+
+  private ExecuteWatchdog watchDog;
+  private IPythonClient ipythonClient;
+  private GatewayServer gatewayServer;
+  private PythonZeppelinContext zeppelinContext;
+  private String pythonExecutable;
+  private String pythonVersion;
+  private long ipythonLaunchTimeout;
+
+  public IPythonInterpreter(Properties properties) {
+    super(properties);
+  }
+
+  @Override
+  public void open() {
+    try {
+      pythonExecutable = getProperty().getProperty("zeppelin.python", "python");
+      ipythonLaunchTimeout = Long.parseLong(
+          getProperty().getProperty("zeppelin.ipython.timeout", "10000"));
+      this.zeppelinContext = new PythonZeppelinContext(
+          null, Integer.parseInt(getProperty().getProperty("zeppelin.python.maxResult", "1000")));
+      int ipythonPort = RemoteInterpreterUtils.findRandomAvailablePortOnAllLocalInterfaces();
+      int jvmGatewayPort = RemoteInterpreterUtils.findRandomAvailablePortOnAllLocalInterfaces();
+      LOGGER.info("Launching IPython Kernel at port: " + ipythonPort);
+      LOGGER.info("Launching JVM Gateway at port: " + jvmGatewayPort);
+      ipythonClient = new IPythonClient("127.0.0.1", ipythonPort);
+      launchIPythonKernel(ipythonPort);
```

Comment:
`launchIPythonKernel` needs `iPythonClient` to check its status. IPython constructor won't connect to that host:port, it would only connect when grpc method is called.